src/cpp/core/file_lock/LinkBasedFileLock.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ * LinkBasedFileLock.cpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <errno.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <set>
+#include <vector>
+
+#include <core/Algorithm.hpp>
+#include <core/Thread.hpp>
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/FilePath.hpp>
+#include <core/FileSerializer.hpp>
+
+#include <boost/foreach.hpp>
+
+namespace rstudio {
+namespace core {
+
+namespace {
+
+std::string makePidString()
+{
+   // get the pid (as a string)
+   std::stringstream ss;
+   ss << (long) ::getpid();
+   return ss.str();
+}
+const std::string& pidString()
+{
+   static std::string instance = makePidString();
+   return instance;
+}
+
+std::string makeHostName()
+{
+   char buffer[256];
+   int status = ::gethostname(buffer, 255);
+   if (status)
+      LOG_ERROR(systemError(errno, ERROR_LOCATION));
+   return std::string(buffer);
+}
+const std::string& hostName()
+{
+   static std::string instance = makeHostName();
+   return instance;
+}
+
+std::string makeProxyLockFileName()
+{
+   return std::string() +
+         ".rstudio-lock" + "-" +
+         hostName() + "-" +
```

Comment:
My understanding is that there are two files involved when using `link` for locking. One needs to be guaranteed-unique, which is what `hostname-pid` gives us -- but since there are threads involved, we may want `hostname-pid-threadid`. The second file is a shared filename that is the other end of the `link`.